### PR TITLE
refactor: ECS7 UI and UiTransform

### PR DESCRIPTION
**Pull request by pbosio**
`CanvasPainter`  was duplicating de UI every time an UI element was updated.
so we took the opportunity to refactor how UI was working to simplify the code.

*`CanvasPainter` was replaced by `ECSScenesUiSystem` system
*`UIText` was temporally removed and will be added on a following PR
*`UITransform` was refactored

the new approach goes as follow:
* When `UITransform` is received from the scene it will create or update internal component `InternalUiContainer`
* `ECSScenesUiSystem` will be in charge in parenting all orphan `InternalUiContainer` since it's possible to try to parent it to an `InternalUiContainer` that does not exist yet
* `ECSScenesUiSystem` will detect scene change to hide previous scene's UI and show (if exists) the UI for the scene where the character is on
* `ECSScenesUiSystem` will create a scene's root UI element (`InternalUiContainer` component for "root scene entity") the first time it detects that the scene has created a new `InternalUiContainer` for any of it's entities

### CODE REVIEW

File: unity-renderer/Assets/DCLPlugins/ECS7/Systems/ScenesUiSystem/ECSScenesUiSystem.cs [hunk 1]
```diff
@@ -0,0 +1,158 @@
...
+                internalUiContainerComponent = internalUiContainerComponent,
+                worldState = worldState,
+                loadedScenes = loadedScenes,
+                lastSceneId = null,
+                isPendingSceneUI = false,
+                currentScene = null
+            };
+
+            return () => Update(state);
+        }
```
[the remark this answers is not in the document]
Comment (AjimenezDCL) [reply unity-renderer/Assets/DCLPlugins/ECS7/Systems/ScenesUiSystem/ECSScenesUiSystem.cs:42]: I also took the time of creating a benchmark project for this. This was the result

```
// * Summary *

BenchmarkDotNet=v0.13.2, OS=Windows 10 (10.0.19044.2006/21H2/November2021Update)
AMD Ryzen 7 3700X, 1 CPU, 16 logical and 8 physical cores
.NET SDK=6.0.401
  [Host]     : .NET 6.0.9 (6.0.922.41905), X64 RyuJIT AVX2
  DefaultJob : .NET 6.0.9 (6.0.922.41905), X64 RyuJIT AVX2


|                     Method |     Mean |    Error |   StdDev |   Gen0 | Allocated |
|--------------------------- |---------:|---------:|---------:|-------:|----------:|
|           WithActionHolder | 18.24 ns | 0.388 ns | 0.580 ns | 0.0057 |      48 B |
|            WithClassHolder | 16.79 ns | 0.271 ns | 0.240 ns | 0.0057 |      48 B |
| WithActionHolderInstancing | 35.29 ns | 0.549 ns | 0.487 ns | 0.0229 |     192 B |
|  WithClassHolderInstancing | 30.11 ns | 0.488 ns | 0.433 ns | 0.0172 |     144 B |
```
`WithActionHolder` use the same approach as the original in the PR. The `action` is retrieved once and reused on every sample.
`WithClassHolder` use the approach above. Only one instance is created and reused on every sample.
`WithActionHolderInstancing` gets the Action on every sample.
`WithClassHolderInstancing` instances a class on every sample.

**Review by AjimenezDCL — CHANGES_REQUESTED** [review]

File: unity-renderer/Assets/DCLPlugins/ECS7/InternalECSComponents/Tests/UiContainerHandlerShould.cs [hunk 2]
```diff
@@ -0,0 +1,73 @@
+using DCL.ECS7.InternalComponents;
+using DCL.ECSRuntime;
+using NUnit.Framework;
+using UnityEngine.UIElements;
+
+namespace Tests
```
Comment (lorux0) [inline unity-renderer/Assets/DCLPlugins/ECS7/InternalECSComponents/Tests/UiContainerHandlerShould.cs:6]: a more specific namespace would be better, for example: `DCL.ECSComponents`

File: unity-renderer/Assets/DCLPlugins/ECS7/Systems/ECSSystemsController.cs [hunk 3]
```diff
@@ -32,6 +35,11 @@ public ECSSystemsController(ECS7System componentWriteSystem, SystemsContext cont
         hoverCanvas.name = "_ECSInteractionHoverCanvas";
         IECSInteractionHoverCanvas interactionHoverCanvas = hoverCanvas.GetComponent<IECSInteractionHoverCanvas>();
 
+        var scenesUiResource = Resources.Load<UIDocument>("ScenesUI");
```
Comment (lorux0) [inline unity-renderer/Assets/DCLPlugins/ECS7/Systems/ECSSystemsController.cs:38]: you can make use of better dependency injection here:
- Environment.i
- DataStore.i
- ECSInteractionHoverCanvas object
-ScenesUI object

File: unity-renderer/Assets/DCLPlugins/ECS7/Systems/ScenesUiSystem/ECSScenesUiSystem.cs [hunk 4]
```diff
@@ -0,0 +1,158 @@
...
+
+                // skip containers with parent already set
+                if (uiContainerData.model.parentElement != null)
+                    continue;
+
+                InternalUiContainer parentDataModel =
+                    internalUiContainerComponent.GetFor(uiContainerData.scene, uiContainerData.model.parentId)?.model;
+
+                // create root entity ui container if needed
+                if (parentDataModel == null && uiContainerData.model.parentId == SpecialEntityId.SCENE_ROOT_ENTITY)
```
Comment (lorux0) [inline unity-renderer/Assets/DCLPlugins/ECS7/Systems/ScenesUiSystem/ECSScenesUiSystem.cs:97]: you can extract to a method and remove comment: `GetOrCreateRootUiContainer`

File: unity-renderer/Assets/DCLPlugins/ECS7/Systems/ScenesUiSystem/ECSScenesUiSystem.cs (hunk 1 above)
[the remark this answers is not in the document]
Comment (pbosio) [reply unity-renderer/Assets/DCLPlugins/ECS7/Systems/ScenesUiSystem/ECSScenesUiSystem.cs]: applied feedback here: https://github.com/decentraland/unity-renderer/pull/3179/commits/fe5a7b935053349e5d2e20a3067c50e1d227f1a7

File: unity-renderer/Assets/DCLPlugins/ECS7/InternalECSComponents/Tests/UiContainerHandlerShould.cs (hunk 2 above)
Comment (pbosio) [reply unity-renderer/Assets/DCLPlugins/ECS7/InternalECSComponents/Tests/UiContainerHandlerShould.cs:6]: The main issue of using a namespace scheme like that for tests is that on unity editor will look like:
```
-> Assambly name
--> DCL
---> ECSComponents
----> Test class name
-----> Test method names
-----> [...] 
```
which makes super hard and a total waste of time when you need to manually select which tests to run, specially when different tests must be selected at the same time

File: unity-renderer/Assets/DCLPlugins/ECS7/Systems/ScenesUiSystem/ECSScenesUiSystem.cs [hunk 5]
```diff
@@ -0,0 +1,158 @@
...
+
+        private static void Update(State state)
+        {
+            string currentSceneId = state.worldState.GetCurrentSceneId();
+            bool sceneChanged = state.lastSceneId != currentSceneId;
+            state.lastSceneId = currentSceneId;
+
+            ApplyParenting(state.internalUiContainerComponent);
+
+            // clear UI if scene changed
```
[the remark this answers is not in the document]
Comment (pbosio) [reply unity-renderer/Assets/DCLPlugins/ECS7/Systems/ScenesUiSystem/ECSScenesUiSystem.cs:59]: Not sure if I understand why having comments is a bad thing

File: unity-renderer/Assets/DCLPlugins/ECS7/Systems/ScenesUiSystem/ECSScenesUiSystem.cs [hunk 6]
```diff
@@ -0,0 +1,158 @@
...
+            if (sceneChanged)
+            {
+                ClearUI(state.uiDocument);
+                state.isPendingSceneUI = !string.IsNullOrEmpty(currentSceneId);
+                state.currentScene = null;
+            }
+
+            // UI not set for current scene yet
+            if (state.isPendingSceneUI)
```
[the remark this answers is not in the document]
Comment (pbosio) [reply unity-renderer/Assets/DCLPlugins/ECS7/Systems/ScenesUiSystem/ECSScenesUiSystem.cs:68]: I find it far more easy to read, understand and debug having the code written like this
```
if (shouldDoSomething)
    DoSomething();
```
than like this
```
void DoSomething(){
    if (!shouldDoSomething)
        return;
    ...
}
```

File: unity-renderer/Assets/DCLPlugins/ECS7/Systems/ScenesUiSystem/ECSScenesUiSystem.cs (hunk 4 above)
Comment (pbosio) [reply unity-renderer/Assets/DCLPlugins/ECS7/Systems/ScenesUiSystem/ECSScenesUiSystem.cs]: applied suggestion here: https://github.com/decentraland/unity-renderer/pull/3179/commits/223c02f15f101e6c89ae7a6a0dd72d3d72038fa9

**Review by AjimenezDCL — APPROVED** [review]

**Pull request merged**